src/loader/utils/load_batch.php

```diff
@@ -28,10 +28,23 @@ function load_batch(array $config, array $transformedevents, callable $loader) {
         $loadedevents = construct_loaded_events($transformedevents, true);
         return $loadedevents;
     } catch (\Exception $e) {
+        $batchsize = count($transformedevents);
         $logerror = $config['log_error'];
-        $logerror("Failed load for event id #" . $eventobj->id . ": " .  $e->getMessage());
+        $logerror("Failed load batch (" . $batchsize . " events)" .  $e->getMessage());
         $logerror($e->getTraceAsString());
-        $loadedevents = construct_loaded_events($transformedevents, false);
+
+        // Recursively retry sending statements in increasingly smaller batches so that only the actual bad data fails.
```

Comment:
I think this will cause some issues because this will delay the time the CRON takes to execute so the next CRON will try to process the same events so we'll duplicate statements. Hope that makes sense.

Reply:
I believe that Moodle cron has some built in stuff that will prevent that happening. See "running cron on multiple servers" here: https://docs.moodle.org/36/en/Cron

Reply:
Ah didn't know that, good spot.
> Tasks can run in parallel and processes use locking to prevent tasks from running at the same time which allows cron to be triggered from multiple web servers that serve the same Moodle instance.

Do you think it would be better to just validate the statements before we send them? Otherwise if the first statement in the batch is invalid we might send X requests where X is the max statements in a batch.

Reply:
Validating before we send them is good, but it'll be a lot of effort to ensure validation is as good and matches the validation done by the LRS. 

X requests would not be sent in that scenario. Let's say it's a batch of 10 statements the first statement is bad. 

Round 1: 1 batch of 10 statements fails
Round 2: 2 batches of 5 statements (1 fails, 1 succeeds)
Round 3: 2 batches of up to 3 statements (1 fails, 1 succeeds)
Round 4: 2 batches of up to 2 statements (1 fails, 1 succeeds)
Round 5: 2 batches of 1 statement (1 fails, 1 succeeds)

Total requests: 9. Ok, that's actually pretty close to 10! 

I'm open to suggestions of a better approach. The current approach of failing the whole batch is not great.

Reply:
Yeah matching the validation of the LRS is pretty much impossible given the slight differences between the LRSs and the little holes in the conformance tests. I'd be interested to know what other people think about this issue (@davidpesce and @caperneoignis). I don't think the current approach is too bad, agree that it's not ideal though, but I'm not sure what the ideal solution is either.

Reply:
I'll also add that even if we get the validation perfect, there's still the odd occasion where there's internet gremlins or an LRS has performance struggles (not our LRSs of course!) and a request fails through no fault of the statements. In that case re-sending is a good thing to do too (and we'd only see 2 extra requests in that scenario).

Reply:
I'd be happy to look into making this a setting if people would feel more comfortable with the option to toggle this behaviour.

Reply:
Hmm... I do like the idea of weeding out the bad statements so the rest of the statements succeeds. Because nothing is more frustrating then see a bunch of statements missing, and find it it was because one statement had an id wrong or html was in the string, or whatever failure. 

I need to look further into it to see what improvements can be made. I'm more on the "we need to improve error reporting and throws" so we can find issues faster. But also really like the idea of a robust batch send. 

And yes @garemoko is correct, the cli locking is quite nice. But if the item in question gets in a infinity loop that could cause some massive lag in cron, which could cause other issues, just a warning.

Reply:
There shouldn't be any infinite loop with this.

Reply:
Just did some maths and for a batch size of 500 that contains 1 bad statement, it's only 18 additional requests. I'm not sure what volumes of statements are realistic for a regularly running cron on an active Moodle site.

Reply:
@garemoko that impressive. That's 3.6% additional request but will allow you to get the 499 other events into the LRS, that is acceptable trade-off. My site generates 5 events from the time a user's login succeeds into the site till they get to the first activitiy. After that, the average is based on what the user is doing. We have a 15 question pre-quiz( 1 event per answered question plus one event for completion, viewing per page, and starting quiz) and a 30 question post quiz, with a tincan_launch activity which will generate about 1.2 event statements per page. So if the user does all three activities, we can be seeing some where in the neighborhood of about 300 events, over 1 to 2 hours. Divide that down to a 5  minute stretch and you are only seeing maybe 5 events, per user, per time block average. 

However, increase that to 1000 concurrent users, and we are seeing about 5K statements per send, give or take. 

I'd almost want to see if we can build in a safe guard for network timeouts, and allow those events to get placed back into the queue and the job just die till next run where it will try the connection again. The other worry, is the job hanging due to a connection timeout or something else, I'd want to put a timeout with a try catch block, Because if a job hangs... I'm not remembering if Moodle handles that gracefully or not or if cron just kind of gets jumbled up and causes a traffic jam. 

FYI: we have our cron running in two minute blocks, with certain jobs schedule for longer windows. But Moodle CRON does allow for locking, so one server will run one thing, the other server will run the other jobs, not locked.

Of course we do our send in real time, because all systems are co-located in the same cloud, so bandwidth is a non-problem and fast. So it's different for my situation.

Reply:
One clarification - the plugin batches events which may generate one or multiple statements. This process is re-running events, not statements, so if the event contains multiple statements and 1 is bad the whole event will still fail. I believe this applies to quiz questions. 

We could make changes to have the task exit in the event of 504 or 503 error codes and send the data later. In that case, you wouldn't want the events to move to the failed events table, you'd just want to leave them where they are. I am cautious about that though as often you would get the 504 and 503 error codes during high traffic periods and not all the requests would fail. Really that's a separate issue anyway because you would want to do that regardless of whether you're re-sending failed requests. I'll create an issue.

Reply:
I made changes so the recursive retry functionality only triggers in the event of a 400 Bad Request response from the LRS. This ensures that in the event of slow responses we're not retrying and retrying so that cron gets blocked. 

I do still think we should retry in the event of 404, 429, 503 and 504, but that can be handled as a separate issue, and probably means retrying next time cron runs rather than in the same cron task.

I like the approach of us tackling specific error codes or groups of error codes one by one, rather than having a catch all approach, so happy for this to start with just 400 errors.

Reply:
Sorry avoided checking in here to avoid getting in trouble with the better half. I'm quite happy to retry on any error code and perhaps only put events in the failed log if we get a 400-499 error code.

Reply:
Also putting this behind a flag would be good I think.

Reply:
I added a setting but haven't tested it just yet. I'll comment again once I have tested. (It's passing automated tests)

Reply:
Cool thanks Andrew 👍